File: java/lesson 10/HowSumRecursive.java
import java.util.*;

class Solution {

    public List<Integer> howSum(int targetSum, int[] numbers) {        
        List<Integer>[] dp = (List<Integer>[]) new ArrayList[targetSum + 1];
        return howSum(targetSum, numbers, dp);
    }

    public List<Integer> howSum(int targetSum, int[] numbers,  List<Integer>[] dp) {
        if (targetSum < 0) return null;
        if (targetSum == 0) return new ArrayList<Integer>();
        if (dp[targetSum] != null) return dp[targetSum];        
        
        for (int number : numbers) {
            List<Integer> current = howSum(targetSum - number, numbers, dp);
            if (current != null) {
                current.add(number);
                dp[targetSum] = current;
                return dp[targetSum];
            }
        }
        
        return null;
    }
}

File: csharp/lesson 10/HowSumRecursive.cs
class Solution
{
    public IList<int> HowSum(int targetSum, int[] numbers)
    {
        var dp = new List<int>[targetSum + 1];
        return HowSum(targetSum, numbers, dp);
    }

    public IList<int> HowSum(int targetSum, int[] numbers, IList<int>[] dp)
    {
        if (targetSum < 0) return null;
        if (targetSum == 0) return new List<int>();
        if (dp[targetSum] != null) return dp[targetSum];

        foreach (int number in numbers)
        {
            var current = HowSum(targetSum - number, numbers, dp);
            if (current != null)
            {
                current.Add (number);
                dp[targetSum] = current;
                return dp[targetSum];
            }
        }

        return null;
    }
}
